foedus-core/include/foedus/xct/xct_id.hpp

```diff
@@ -165,6 +165,113 @@ struct McsBlock {
   }
 };
 
+struct McsRwBlock {
+  static const uint8_t kStateClassMask       = 3U;        // [LSB + 1, LSB + 2]
+  static const uint8_t kStateClassReaderFlag = 1U;        // LSB binary = 01
+  static const uint8_t kStateClassWriterFlag = 2U;        // LSB binary = 10
+
+  static const uint8_t kStateBlockedFlag     = 1U << 7U;  // MSB binary = 1
+  static const uint8_t kStateBlockedMask     = 1U << 7U;
+
+  static const uint8_t kSuccessorClassReader = 1U;
+  static const uint8_t kSuccessorClassWriter = 2U;
+  static const uint8_t kSuccessorClassNone   = 3U;        // LSB binary 11
+
+  union Self {
+    uint16_t data_;                       // +2 => 2
+    struct Components {
+      uint8_t successor_class_;
+      // state_ covers:
+      // Bit 0-1: my **own** class (am I a reader or writer?)
+      // Bit 7: blocked (am I waiting for the lock or acquired?)
+      uint8_t state_;
+    } components_;
+  } self_;
+  // TODO(tzwang): make these two fields 8 bytes by themselves. Now we need
+  // to worry about sub-word writes (ie have to use atomic ops even when
+  // changing only these two fields because they are in the same byte as data_).
+  thread::ThreadId successor_thread_id_;  // +2 => 4
+  McsBlockIndex successor_block_index_;   // +4 => 8
+
```

Comment:
Let's make this 16 bytes in near future.

uint32_t for successor_thread_id (maybe you want to embed successor_class here)
uint32_t for successor_block_index
uint64_t other-stuffs

First 8 bytes are written only by successor.
Second 8 bytes needs a bit more thinking due for RW-MCSg.